File: svc-md/src/main/java/com/beancounter/marketdata/providers/fxrates/FxGateway.java
package com.beancounter.marketdata.providers.fxrates;

import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

/**
 * Api calls to exchangeratesapi.io.
 *
 * @author mikeh
 * @since 2019-03-03
 */
@Configuration
@FeignClient(
    name = "fxRequest",
    url = "${beancounter.marketdata.provider.fx.url:https://api.exchangeratesapi.io}")

public interface FxGateway {
  @RequestMapping(
      method = RequestMethod.GET,
      headers = {"Content-Type: text/plain"},
      value = "/{date}?base={base}&symbols={symbols}&"
  )
  EcbRates getRatesForSymbols(
      @PathVariable("date") String date,
      @PathVariable("base") String base,
      @PathVariable("symbols") String symbols);


}

File: svc-data/src/main/kotlin/com/beancounter/marketdata/fx/fxrates/FxGateway.kt
package com.beancounter.marketdata.fx.fxrates

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.client.RestClient
import org.springframework.web.client.body

/**
 * API calls to exchangeratesapi.io using RestClient.
 */
@Component
class FxGateway(
    @Qualifier("exchangeRatesRestClient")
    private val restClient: RestClient,
    @Value($$"${beancounter.market.providers.fx.key:}")
    private val accessKey: String
) {
    fun getRatesForSymbols(
        date: String,
        base: String,
        symbols: String
    ): ExRatesResponse? =
        restClient
            .get()
            .uri(
                "/v1/{date}?base={base}&symbols={symbols}&access_key={accessKey}",
                date,
                base,
                symbols,
                accessKey
            ).retrieve()
            .body<ExRatesResponse>()
}